platform/mds/mds/src/main/java/org/motechproject/mds/lookup/LookupExecutor.java

```diff
@@ -61,8 +66,22 @@ public Object execute(Map<String, ?> lookupMap, QueryParams queryParams) {
             return MethodUtils.invokeMethod(dataService, lookup.getMethodName(),
                     args.toArray(new Object[args.size()]),
                     argTypes.toArray(new Class[argTypes.size()]));
-        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
-            throw new LookupExecutorException("Unable to execute lookup " + lookup.getLookupName(), e);
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+            throw new LookupExecutorException(lookupExceptionMessage + lookup.getLookupName() + ".", e, null);
+        } catch (InvocationTargetException e) {
+            if (e.getTargetException() instanceof JDOUserException) {
+                JDOUserException userException = (JDOUserException) e.getTargetException();
+                lookupExceptionMessageKey = "mds.error.lookupExecUserError";
+                for (Throwable exception : userException.getNestedExceptions()) {
+                    if (exception instanceof QueryNotUniqueException) {
+                        lookupExceptionMessageKey = "mds.error.lookupExecNotUniqueError";
+                        throw new LookupExecutorException(lookupExceptionMessage + lookup.getLookupName() + ".", e, lookupExceptionMessageKey);
+                    }
+                }
+                throw new LookupExecutorException(lookupExceptionMessage + lookup.getLookupName() + ".", e, lookupExceptionMessageKey);
+            } else {
+                throw new LookupExecutorException(lookupExceptionMessage + lookup.getLookupName() + ".", e, lookupExceptionMessageKey);
```

Comment:
From what I can see, all these throws look exactly the same. Only the message differs. Let's move all of those throws outside the if blocks and only set the message keys there.